include/zephyr/senss/senss.h

```diff
@@ -0,0 +1,522 @@
+ * @defgroup senss_physical Physical Sensor API
+ * @ingroup senss_sensor
+ * @defgroup senss_virtual Virtual Sensor API
+ * @ingroup senss_sensor
+ */
+
+#include <senss/senss_datatypes.h>
+#include <senss/senss_sensor_types.h>
+#include <senss/senss_sensor.h>
+
+/**
+ * @brief Sensor Subsystem API
+ * @addtogroup senss_api
+ * @{
+ */
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Initialize sensor subsystem core.
+ *
+ * This will allocate internal resources, such as create internal schedule
+ * task, create internal management data structures.
+ *
+ * It will also enumerate all sensor instances which registered by
+ * \ref SENSS_REGISTER_SENSOR, build sensors report relational data
+ * model, if a virtual sensor's required dependency sensor type not found, will
+ * return error.
+ * Then call each sensor instances' init callback API.
+ *
+ * @return 0 on success or negative error value on failure.
+ *
+ */
+int senss_init(void);
+
+ /**
+  * @brief De-initialize sensor subsystem core.
+  *
+  * This will release all dynamically allocated internal resources, and call
+  * all sensor instances' deinit callback API.
+  *
+  * @return 0 on success or negative error value on failure.
+  */
+int senss_deinit(void);
+
+ /**
+  * @brief Find sensor instances by sensor type and instance ID.
+  *
+  * Application can use these API to find sensor instances information by
+  * sensor types and instance id, or enumerate all sensor instances'
+  * information at one call.
+  *
+  * This API just returned read only information of sensor instances,
+  * no side effect to sensor instances.
+  *
+  * @param type The sensor type which need to find,
+  * \ref SENSS_SENSOR_TYPE_ALL for all types.
+  *
+  * @param instance The instance ID which need to find, -1 for find all
+  * instances with type \p type
+  *
+  * @param max The max count of the \p infos array input. Can get real count
+  * number via \ref senss_get_sensor_count.
+  *
+  * @param *info Input info array for receiving found sensor instances
+  *
+  * @return Number of sensor instances found, 0 returned if no found
+  */
+int senss_find_sensor(int type, int instance,
+		 int max, struct senss_sensor_info *info);
+
+ /**
+  * @brief Get sensor instance count of given type and instance ID.
+  *
+  * @param type The sensor type for checking, \ref SENSS_SENSOR_TYPE_ALL
+  * for all types.
+  *
+  * @param instance The instance ID for checking, -1 for all instances with
+  * \p type type
+  *
+  * @return Count of matched sensor instances, 0 returned if no matched.
+  */
+int senss_get_sensor_count(int type, int instance);
+
+/**
+ * @brief Open sensor instance by sensor type and instance ID.
+ *
+ * Application clients use it to open a sensor instance and get its handle.
+ * Support multiple Application clients for open same sensor instance,
+ * in this case, the returned handle will different for different clients.
+ *
+ * @param type The sensor type which need to find.
```

Comment:
We already found the sensor since we have the |instance| of the sensor to open. Is |type| necessary?

Reply:
`instance ID` is not global,  but under `type`,  for example,  with same accelerometer sensor type, for multiple sensor instances, the  instance ID counted from 0, 1, 2...  And for gyroscope sensor type multiple sensor instances,  the `instance ID` also counted from 0,1, 2...

Reply:
`instance_id` is confusing. CHRE calls this `sensor_index` which is a bit more intuitive. I also thought this was an absolute ID similar to the handle.

Reply:
Yes,  `sensor_index` is more intuitive, now changed to `sensor_index`,  thanks a lot!